onnxruntime/core/graph/contrib_ops/contrib_defs.cc

```diff
@@ -904,6 +905,194 @@
                                   updateOutputElemType(ctx, 0, ONNX_NAMESPACE::TensorProto::BOOL);
                                 }));
 
+void col2imShapeInference(InferenceContext& ctx) {
+  propagateElemTypeFromInputToOutput(ctx, 0, 0);
+
+  // All inputs shapes are required
+  if (!hasNInputShapes(ctx, 3)) {
+    return;
+  }
+
+  // Assuming image_shape has correct spatial dimensions and reused for next validation steps
+  //       An alternative is get the the number of spatial dimensions as an input
+  if (ctx.getInputType(1)->tensor_type().shape().dim_size() != 1) {
+    fail_shape_inference("image_shape tensor must have rank 1.");
+  }
+  size_t n_input_dims = static_cast<size_t>(ctx.getInputType(1)->tensor_type().shape().dim(0).dim_value());
+  std::vector<int64_t> image_shape = {};
+  const TensorProto* image_shape_data = ctx.getInputData(1);
+  if (image_shape_data) {
+    image_shape = ParseData<int64_t>(image_shape_data);
+    if (image_shape.size() != n_input_dims) {
+      fail_shape_inference("image_shape tensor must have ", n_input_dims, " spatial dimensions.");
+    }
+  }
+
+  std::vector<int64_t> pads = {};
+  if (getRepeatedAttribute(ctx, "pads", pads)) {
+    if ((pads.size() != 0) && (pads.size() != n_input_dims * 2)) {
+      fail_shape_inference("Attribute pads has incorrect size");
+    }
+  }
+
+  std::vector<int64_t> dilations = {};
+  if (getRepeatedAttribute(ctx, "dilations", dilations)) {
+    if ((dilations.size() != 0) && (dilations.size() != n_input_dims)) {
+      fail_shape_inference("Attribute dilations has incorrect size");
+    }
+  }
+
+  std::vector<int64_t> strides = {};
+  if (getRepeatedAttribute(ctx, "strides", strides)) {
+    if ((strides.size() != 0) && (strides.size() != n_input_dims)) {
+      fail_shape_inference("Attribute strides has incorrect size");
+    }
+  }
+
+  auto input_shape = ctx.getInputType(0)->tensor_type().shape();
+  if (input_shape.dim_size() != 3) {
+    fail_shape_inference("input must have rank 3.");
+  }
+
+  std::vector<int64_t> block_shape = {};
+  const TensorProto* block_shape_data = ctx.getInputData(2);
+  if (block_shape_data) {
+    block_shape = ParseData<int64_t>(block_shape_data);
+    if (block_shape.size() != n_input_dims) {
+      fail_shape_inference("block_shape tensor must have ", n_input_dims, " spatial dimensions.");
+    }
+  }
+  if (ctx.getInputType(2)->tensor_type().shape().dim_size() != 1) {
+    fail_shape_inference("block_shape tensor must have rank 1.");
+  } else if (
+      (ctx.getInputType(2)->tensor_type().shape().dim(0).has_dim_value()) &&
```

Comment:
>has_dim_value( [](http://example.com/codeflow?start=57&length=14)

utils::HasDimValue()

Reply:
For my understanding, what is the reason for `utils::HasDimValue()` as opposed to the ONNX `has_dim_value()` API?

Reply:
These utils were introduced to hide the difference between proto2 and proto3. The latter does not have has_*() interfaces or at least was not planned to have a year ago. So we introduced inline helpers to deal with that.